pkg/remotecluster/remotecluster.go

```diff
@@ -0,0 +1,223 @@
+
+	resourceWatcher := &RemoteCluster{
+		stopCh:                make(chan struct{}),
+		ClusterID:             clusterID,
+		ClientSet:             clientSet,
+		podInformer:           podInformer,
+		networkPolicyInformer: networkPolicyInformer,
+		eventChanMutex:        &sync.Mutex{},
+	}
+
+	podInformer.AddEventHandler(resourceWatcher)
+	networkPolicyInformer.AddEventHandler(resourceWatcher)
+
+	return resourceWatcher
+}
+
+func (rc *RemoteCluster) HasSynced() bool {
+	return rc.podInformer.HasSynced() &&
+		rc.networkPolicyInformer.HasSynced()
+}
+
+// Stop will stop the running informers
+func (rc *RemoteCluster) Stop() {
+	close(rc.stopCh)
+}
+
+func (rc *RemoteCluster) Stopped() bool {
+	select {
+	case _, ok := <-rc.stopCh:
+		// channel is not ok to read, means it's closed now
+		return !ok
+	default:
+		// channel was ok to read, but had no messages
+		return false
+	}
+}
+
+func (rc *RemoteCluster) Run(onSyncDoneFunc func(resourceWatcher *RemoteCluster)) {
+	go rc.podInformer.Run(rc.stopCh)
+	go rc.networkPolicyInformer.Run(rc.stopCh)
+
+	go func() {
+		if !cache.WaitForCacheSync(rc.stopCh, rc.podInformer.HasSynced) {
+			klog.Warning("Timed out waiting for pod informer to sync")
+		}
+
+		if !cache.WaitForCacheSync(rc.stopCh, rc.networkPolicyInformer.HasSynced) {
+			klog.Warning("Timed out waiting for NetworkPolicy informer to sync")
+		}
+
+		if onSyncDoneFunc != nil {
+			onSyncDoneFunc(rc)
+		}
+	}()
+}
+
+func (rc *RemoteCluster) GetPods() []interface{} {
+	return rc.podInformer.GetStore().List()
+}
+
+func (rc *RemoteCluster) GetNetworkPolicies() []interface{} {
+	return rc.networkPolicyInformer.GetStore().List()
+}
+
+func (rc *RemoteCluster) SetEventChannel(eventChan chan *Event) {
+	rc.eventChanMutex.Lock()
+	rc.eventChan = eventChan
+	rc.eventChanMutex.Unlock()
+}
+
+func (rc *RemoteCluster) OnAdd(obj interface{}) {
+	rc.enqueueEvent(rc.NewAddEvent(obj))
+}
+
+func (rc *RemoteCluster) OnDelete(obj interface{}) {
+	rc.enqueueEvent(rc.NewDeleteEvent(obj))
+}
+
+func (rc *RemoteCluster) OnUpdate(oldObj interface{}, newObj interface{}) {
+	rc.enqueueEvent(rc.NewUpdateEvent(oldObj, newObj))
+}
+
+func (rc *RemoteCluster) enqueueEvent(event *Event) {
+
+	// lock used as memory barrier to make sure eventChannel ref
+	// is synchronized between threads
+	rc.eventChanMutex.Lock()
+	defer rc.eventChanMutex.Unlock()
+	if rc.eventChan != nil {
+		rc.eventChan <- event
+	}
+
+}
+
+func (rc *RemoteCluster) NewAddEvent(objInterface interface{}) *Event {
+
+	event := Event{
+		Cluster: rc,
+		Type:    AddEvent,
+		Objs:    []interface{}{objInterface},
+	}
+
+	return rc.extractEventDetails(objInterface, &event)
+}
+
+func (rc *RemoteCluster) NewUpdateEvent(objInterface, newObjInterface interface{}) *Event {
+
+	event := Event{
+		Cluster: rc,
+		Type:    UpdateEvent,
+		Objs:    []interface{}{objInterface, newObjInterface},
+	}
+	return rc.extractEventDetails(newObjInterface, &event)
```

Comment:
I find it confusing that sometimes you use `event` and sometimes `&event` throughout the code